src/zookeeperLock.ts

```diff
@@ -122,32 +122,56 @@ export class ZookeeperLock {
         return this.connect(this.config.spinDelay);
     };
 
-    public lock = (key : string) : Promise<any> => {
+    public lock = (key : string, timeout? : number) : Promise<any> => {
         var path = `/locks/${this.config.pathPrefix ? this.config.pathPrefix + '/' : '' }`;
         var nodePath = `${path}${key}`;
 
         debuglog(`try locking ${key} at ${path}`);
 
-        return Promise<any>((resolve, reject) => {
+         var p = Promise<any>((resolve, reject) => {
+             var timedOut = false;
+             if (timeout) {
+                 setTimeout(() => {
+                     timedOut = true;
+                 }, timeout);
+             }
             this.connect()
-                .then(() => {
-                    debuglog(`making lock at ${nodePath}`);
-                    return this.makeLockDir(nodePath);
-                })
-                .then(() => {
-                    return this.initLock(nodePath);
-                })
-                .then(() => {
-                    debuglog(`waiting for lock at ${nodePath}`);
-                    return this.waitForLock(nodePath);
-                })
-                .then((lock : ZookeeperLock) => {
-                    debuglog('lock acquired');
-                    resolve(true);
-                }).catch((err) => {
+            .then(() => {
+                if (timedOut) { throw new Error('timeout'); }
+
+                debuglog(`making lock at ${nodePath}`);
+                return this.makeLockDir(nodePath);
+            })
+            .then(() => {
+                if (timedOut) { throw new Error('timeout'); }
+
+                return this.initLock(nodePath);
+            })
+            .then(() => {
+                if (timedOut) { throw new Error('timeout'); }
+
+                debuglog(`waiting for lock at ${nodePath}`);
+                return this.waitForLock(nodePath);
+            })
+            .then((lock : ZookeeperLock) => {
+                if (timedOut) { throw new Error('timeout'); }
+
+                debuglog('lock acquired');
+                resolve(true);
+            }).catch((err) => {
+                if (timedOut) {
+                    this.disconnect();
+                } else {
                     reject(err);
-                });
+                }
+            });
         });
+
+        if (timeout) {
+            p = p.timeout(timeout);
```

Comment:
not that magic... https://github.com/kriskowal/q/wiki/API-Reference#promisetimeoutms-message